scripts/word_embeddings/tools/cooccur.cc

```diff
@@ -0,0 +1,368 @@
+//  Tool to calculate word-word cooccurrence statistics
+//
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+// * Includes and definitions
+#include <cstdlib>
+#include <filesystem>
+#include <fstream>
+#include <functional>
+#include <iostream>
+#include <numeric>
+#include <queue>
+#include <random>
+#include <set>
+#include <sstream>
+#include <stdexcept>
+#include <vector>
+
+#include "CLI/CLI.hpp"      // command line parser
+#include "cnpy.h"           // numpy
+#include "range/v3/all.hpp" // Ranges TS https://github.com/ericniebler/range-v3/
+#include "sparsepp/spp.h"   // fast sparse hash map
+
+#include "utils.h"
+
+namespace rsv = ranges::view;
+namespace fs = std::filesystem;
+using Vocab = spp::sparse_hash_map<std::string, std::pair<uint32_t, uint32_t>>;
+using count_type = float;
```

Comment:
Is using float safe? float(50M) == float(50M+1) last time i checked with numpy

Reply:
In this case we don't really care if the count is 1 or 2 different. Given that the counts are based on a noisy corpus. So the memory saving of float vs double seems more important.